quinn-proto/src/connection/mod.rs

```diff
@@ -2502,16 +2499,23 @@ impl Connection {
                         }
                     }
 
-                    if self.side.is_server() && self.peer_params.stateless_reset_token.is_none() {
-                        if self.rem_cids.active_seq() == 0 {
-                            // We're a server still using the initial remote CID for the client, so
-                            // let's switch immediately to enable clientside stateless resets.
-                            let _ = self.update_rem_cid();
-                        }
-                    } else if self.rem_cids.is_active_retired() {
-                        // If our current CID is meant to be retired; or
-                        // active remote CID is invalid (due to packet loss or reordering),
-                        // we must switch to next valid CID
+                    if !retired.is_empty() {
```

Comment:
If the `retired.is_empty()` invariant implies that `active_reset_token()` must be `Some`, can we rely on that more directly?

Reply:
This was a very good question and led me on an adventure that ended in refactoring the entire thing again, this time better encapsulating the NEW_CONNECTION_ID handling logic in CidQueue by way of merging `insert` and `retire_prior_to`.